src/rt/ds/bag.h

```diff
@@ -0,0 +1,250 @@
+    {
+      MaybeElem* hole = (MaybeElem*)item;
+      hole->hole_ptr = (MaybeElem*)((uintptr_t)next_free | EMPTY_MASK);
+      next_free = hole;
+    }
+
+    /// returns true if this bag is empty
+    ALWAYSINLINE bool empty()
+    {
+      return index == null_index;
+    }
+
+    /// Return the top element of the bag without removing it.
+    ALWAYSINLINE Elem* peek()
+    {
+      assert(!empty());
+      return &(index->item);
+    }
+
+    /// Insert an element into the bag.
+    ALWAYSINLINE Elem* insert(Elem item, Alloc& alloc)
+    {
+      if (next_free != nullptr)
+      {
+        MaybeElem* prev = next_free->hole_ptr;
+        next_free->item = item;
+        assert((uintptr_t)prev & EMPTY_MASK);
+        MaybeElem* cur = next_free;
+        next_free = (MaybeElem*)((uintptr_t)prev & ~EMPTY_MASK);
+        return &(cur->item);
+      }
+      if (!is_full(index))
+      {
+        index++;
+        index->item = item;
+        return &(index->item);
+      }
+      return insert_slow(item, alloc);
+    }
+
+  private:
+    /// Slow path for insert, performs an insert, when allocation is required.
+    Elem* insert_slow(Elem item, Alloc& alloc)
+    {
+      assert(is_full(index));
+
+      Block* next = (Block*)alloc.template alloc<sizeof(Block)>();
+      assert(((uintptr_t)next) % alignof(Block) == 0);
+      next->prev.hole_ptr = index;
+      index = &(next->data[0]);
+      index->item = item;
+      return &(index->item);
+    }
+
+  public:
+    class iterator
+    {
+      friend class Bag;
+
+    public:
+      iterator(Bag<T, U, Alloc>* bag) : bag(bag)
+      {
+        // If the bag is empty then there is nothing to iterate over.
+        if (bag->empty())
+        {
+          ptr = nullptr;
+          return;
+        }
+
+        ptr = bag->peek();
+      }
+
+      iterator(Bag<T, U, Alloc>* bag, Elem* p) : bag(bag), ptr(p) {}
+
+      iterator operator++()
+      {
+        auto p = (MaybeElem*)ptr;
+        do
+        {
+          p--;
+          if (p->hole_ptr == Bag::null_index)
+          {
+            ptr = nullptr;
+            return *this;
+          }
+
+          if (bag->is_empty(p))
+          {
+            p = bag->get_block(p)->prev.hole_ptr;
+          }
+        } while ((uintptr_t)p->hole_ptr & EMPTY_MASK);
+
+        assert(
+          (p == Bag::null_index) ||
+          (((uintptr_t)p->hole_ptr & EMPTY_MASK) == 0));
+
+        ptr = (Elem*)p;
+        return *this;
+      }
+
+      inline bool operator!=(const iterator& other) const
+      {
+        return ptr != other.ptr;
+      }
+
+      inline bool operator==(const iterator& other) const
+      {
+        return ptr == other.ptr;
+      }
+
+      inline Elem* operator*() const
+      {
+        return ptr;
+      }
+      inline iterator begin()
+      {
+        return {bag};
+      }
+
+      inline iterator end()
+      {
+        return {bag, nullptr};
+      }
+
+    private:
+      Bag<T, U, Alloc>* bag;
+      Elem* ptr;
```

Comment:
I guess given my suggestion this would have to be
```suggestion
      MaybeElem* ptr;
```
```suggestion
      Elem* ptr;
```